File: 2094-remove-stones-to-minimize-the-total/remove-stones-to-minimize-the-total.py
import heapq
import math

class Solution:
    def minStoneSum(self, piles: List[int], k: int) -> int:
        # Create a max heap from the negative values of the piles
        max_heap = [-x for x in piles]
        heapq.heapify(max_heap)
        
        # Perform k operations
        while k:
            # Fetch the maximum value from the heap
            max_val = -heapq.heappop(max_heap)
            # Calculate the new value after operation
            new_val = max_val - math.floor(max_val / 2)
            # Push the new value back to the heap
            heapq.heappush(max_heap, -new_val)
            k -= 1
        
        # Calculate the sum of remaining values in the heap
        return sum(-x for x in max_heap)

File: 2094-remove-stones-to-minimize-the-total/remove-stones-to-minimize-the-total.cpp
#include <queue>
#include <cmath>
#include <vector>

class Solution {
public:
    int minStoneSum(std::vector<int>& piles, int k) {
        // Create a max heap
        std::priority_queue<int> max_heap;
        for (int pile : piles) {
            max_heap.push(pile);
        }
        
        // Perform k operations
        while (k--) {
            // Fetch the maximum value from the heap
            int max_val = max_heap.top();
            max_heap.pop();
            // Calculate the new value after operation
            int new_val = max_val - std::floor(max_val / 2);
            // Push the new value back to the heap
            max_heap.push(new_val);
        }
        
        // Calculate the sum of remaining values in the heap
        int sum = 0;
        while (!max_heap.empty()) {
            sum += max_heap.top();
            max_heap.pop();
        }
        
        return sum;
    }
};
